cmd/logmonitor/db.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian/types"
+	"github.com/jinzhu/gorm"
```

Comment:
Consider defining a storage interface like we do [here](https://github.com/google/trillian/blob/7b07980fdb5a5b64ab920b59e4a1f9dd3275b455/server/storage_provider.go#L90:6)   This would allow the log monitor to be used by other storage implementations such as in-memory fakes, cloud spanner, and amazon's databases.

Reply:
I prefer to use a proper ORM rather than have to roll my own.

Gorm is actually pretty good. I didn't have to write any dialect-specific code, and I get support for MySQL, PostgreSQL, SQLite and Microsoft SQL Server (and any potential new dialects that get added to Gorm in the future) for free. Plus, it was very easy to get the binary to also handle setting up and migrating its own database tables.

Gorm's support for PostgreSQL and MySQL means that this monitor already supports CockroachDB and almost all of Amazon RDS's database engines (including Amazon Aurora), with the only exception being the Oracle RDS database engine.

For in-memory fakes, I would argue that using SQLite (either as a temporary on-disk database or an in-memory one) is a much better option than rolling my own in-memory fake for testing. The tests in `main_test.go` are a good example of this, since they are using a real (but temporary) SQLite database.

As for cloud spanner, given gorm's popularity, I suspect a cloud spanner dialect will eventually get added. I have actually been considering adding that support myself since it's not much work to add a new dialect, but I haven't gotten around to it yet.

Reply:
Our log / map data model isn't complex enough to really benefit from an ORM and we want to control how locks are obtained. Possibly it is for a monitor but it could be arguable. Performance might be an issue as Gorm presumably can't do code generation like Hibernate etc. can do for Java? This might not be an issue if you're monitoring a small number of logs. Hard to say at this point.

Another issue for an ORM is that you're potentially on the hook for supporting configurations you haven't tested. Again, possibly not as vital for a monitor but it was a concern for us.

Testing a server with SQLite is tough and we gave up on it. We had a lot of test flakes. Might not happen with simple unit tests but I think it's worth mentioning.

CloudSpanner isn't just another SQL dialect. The way you interact with transactions changes. If you want to support it then it's worth following their model from the start rather than having to rework it later as we did.